main.go

```diff
@@ -0,0 +1,240 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/urfave/cli"
+)
+
+const (
+	imageListPutImagesPath = "/images"
+)
+
+var (
+	tagRegexp = regexp.MustCompile(`:([\w][\w.-]{0,127})$`)
+	logger    = log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lshortfile)
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "imagelist-docker-events"
+	app.Version = "v0.0.1"
+	app.Usage = "poll docker for image push events and add them to imagelist service"
+
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "imagelist-url",
+			Usage: "imagelist service url",
+		},
+	}
+
+	app.Action = func(c *cli.Context) error {
+		f := filters.NewArgs()
+		f.Add("type", events.ImageEventType)
+
+		if !c.IsSet("imagelist-url") {
+			return cli.NewExitError("error: imagelist-url needs to be set", 1)
+		}
+		u := c.String("imagelist-url")
+		imageListURL, err := joinURL(u, imageListPutImagesPath)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("error: failed to parse imagelist url: %v", err), 1)
+		}
+
+		// main loop
+		for {
+			c, err := client.NewEnvClient()
+			if err != nil {
+				logger.Printf("error: docker connection failed: %v", err)
+			}
+			defer c.Close()
+
+			messages, errs := c.Events(context.Background(), types.EventsOptions{Filters: f})
+			if err := processEvents(messages, errs, imageListURL); err != nil {
+				logger.Print(err)
+				<-time.After(time.Second)
+				continue
+			}
+		}
+
+	}
+
+	app.Run(os.Args)
+}
+
+func processEvents(messages <-chan events.Message, errs <-chan error, url string) error {
+	select {
+	case err := <-errs:
+		if err != nil && err != io.EOF {
+			return fmt.Errorf("error: failed to read events: %v", err)
+		}
+	case m := <-messages:
+		if m.Type == "image" && m.Action == "push" {
+			go addToImageList(m.ID, url)
+		}
+	}
+	return nil
+}
+
+type image struct {
+	ID   string   `json:"id"`
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+func addToImageList(name string, url string) {
+	images, err := getRepoDigests(name)
+	if err != nil {
+		logger.Print(err)
+		return
+	}
+
+	for _, i := range images {
+		go func(i image) {
+			attempt := 0
+			for {
+				if attempt == 30 {
+					logger.Printf("error submitting %s image: max retries reached", name)
+					return
+				}
+
+				if attempt != 0 {
+					<-time.After(time.Second * 3)
```

Comment:
Maybe some form of exponential backoff could be useful here?

Reply:
Also, this probably needs a more "intelligent" retry mechanism – e.g. if it's a JSON parsing issue then it's probably a non-transient bug, so it wouldn't make sense to retry,

I wonder if Golang has an HTTP retry library of some form? So then the HTTP library itself could handle retries transparently?